### PR TITLE
KubeRay node provider refactor

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Implements KubeRay node provider as a "BatchingNodeProvider".
Builds on https://github.com/ray-project/ray/pull/29933.

### Summary of design

An autoscaler update now works like this:

- list pod data from k8s
- check if it's safe to proceed with update. Abort the update if not.
- do some internal calculation to determine desired scale
- submit a single patch to the RayCluster CR if a scale change is required

Everything is single-threaded and there are O(1) K8s API calls per autoscaler update.

A follow-up PR will have the list pods operation specify a resource version, to reduce the load on etcd.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [x] Existing KubeRay autoscaling integration test

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
Needs one more unit test.

**Comment by DmitriGekhtman** [comment]
This is good to review.

File: python/ray/autoscaler/_private/kuberay/node_provider.py [hunk 1]
```diff
@@ -170,31 +208,152 @@ def __init__(
...
-            provider_config.get(FOREGROUND_NODE_LAUNCH_KEY, False) is True
-        ), f"To use KuberayNodeProvider, must set `{FOREGROUND_NODE_LAUNCH_KEY}:True`."
         assert (
             provider_config.get(WORKER_LIVENESS_CHECK_KEY, True) is False
         ), f"To use KuberayNodeProvider, must set `{WORKER_LIVENESS_CHECK_KEY}:False`."
         assert (
             provider_config.get(WORKER_RPC_DRAIN_KEY, False) is True
         ), f"To use KuberayNodeProvider, must set `{WORKER_RPC_DRAIN_KEY}:True`."
-        provider_exists = True
+        BatchingNodeProvider.__init__(
```
Comment (wuisawesome) [inline python/ray/autoscaler/_private/kuberay/node_provider.py:217]: nit I think `__super__(...)` is the preferred way of doing this.

**Review by wuisawesome — APPROVED** [review]

File: python/ray/autoscaler/_private/kuberay/node_provider.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:217]: I've seen both ways in the codebase, but I prefer to make it explicit.

**Comment by DmitriGekhtman** [comment]
Test failures are unrelated.

**Pull request merged**